File: BashX/Services/SpeedTester.swift
import Foundation
import Network

actor SpeedTester {
    struct Result: Sendable {
        let name: String
        let delayMs: Int
    }

    /// Live proxy names from mihomo `/proxies` — used when UI name differs slightly from runtime.
    private var proxyCatalog: [String: String] = [:]
    private var catalogController = ""

    /// Like Clash Verge / Stash: measure via mihomo `GET /proxies/{name}/delay` (URLTest through that node).
    /// API calls always bypass system proxy (`connectionProxyDictionary = [:]`).
    /// TCP handshake is only a last-resort fallback when the core API is unavailable.
    func testAll(
        nodes: [ProxyNode],
        timeoutMs: Int,
        concurrency: Int,
        controller: String? = nil,
        secret: String = "",
        testURL: String = "http://www.gstatic.com/generate_204",
        onProgress: @MainActor @escaping (String, Int) -> Void
    ) async -> [Result] {
        let testables = nodes.filter { ClashConfigParser.isSpeedTestable($0) }
        guard !testables.isEmpty else { return [] }

        let limit = max(1, min(concurrency, 16))
        // Keep per-node budget tight — sequential URL fallbacks used to inflate failures.
        let apiTimeout = max(min(timeoutMs, 8000), 2000)
        let useAPI = !(controller?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true)
        if useAPI, let controller {
            await refreshProxyCatalog(controller: controller, secret: secret)
        }

        var results: [Result] = []
        results.reserveCapacity(testables.count)

        await withTaskGroup(of: Result.self) { group in
            var iterator = testables.makeIterator()

            func enqueueNext() {
                guard let node = iterator.next() else { return }
                group.addTask {
                    let delay: Int
                    if useAPI, let controller {
                        delay = await self.apiDelay(
                            node: node,
                            controller: controller,
                            secret: secret,
                            timeoutMs: apiTimeout,
                            testURL: testURL
                        )
                    } else {
                        delay = await self.tcpDelay(host: node.server, port: node.port, timeoutMs: apiTimeout)
                    }
                    return Result(name: node.name, delayMs: delay)
                }
            }

            for _ in 0..<min(limit, testables.count) {
                enqueueNext()
            }

            for await result in group {
                results.append(result)
                await onProgress(result.name, result.delayMs)
                enqueueNext()
            }
        }

        return results
    }

    private func refreshProxyCatalog(controller: String, secret: String) async {
        if catalogController == controller, !proxyCatalog.isEmpty { return }
        catalogController = controller
        proxyCatalog.removeAll(keepingCapacity: true)

        guard let url = URL(string: "http://\(controller)/proxies") else { return }
        var request = URLRequest(url: url, timeoutInterval: 4)
        applyAuth(&request, secret: secret)
        let config = URLSessionConfiguration.ephemeral
        config.connectionProxyDictionary = [:]
        guard let (data, response) = try? await URLSession(configuration: config).data(for: request),
              let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let proxies = json["proxies"] as? [String: Any] else { return }

        for (name, value) in proxies {
            guard let dict = value as? [String: Any] else { continue }
            let type = (dict["type"] as? String)?.lowercased() ?? ""
            if ["direct", "reject", "select", "url-test", "fallback", "load-balance", "relay", "pass", "compatible"].contains(type) {
                continue
            }
            proxyCatalog[name] = name
            if let server = dict["server"] as? String, let port = intValue(dict["port"]) {
                let key = "\(server.lowercased()):\(port)"
                proxyCatalog[key] = name
            }
        }
    }

    private func resolveProxyName(_ node: ProxyNode) -> String {
        if proxyCatalog[node.name] != nil { return node.name }
        let key = "\(node.server.lowercased()):\(node.port)"
        if let resolved = proxyCatalog[key] { return resolved }
        return node.name
    }

    private func apiDelay(
        node: ProxyNode,
        controller: String,
        secret: String,
        timeoutMs: Int,
        testURL: String
    ) async -> Int {
        let proxyName = resolveProxyName(node)
        let urls = testURLCandidates(primary: testURL)
        for url in urls {
            if let ms = await apiDelayOnce(
                proxyName: proxyName,
                controller: controller,
                secret: secret,
                timeoutMs: timeoutMs,
                testURL: url
            ), ms > 0 {
                return ms
            }
        }
        // Do not invent numbers from history — failed probe stays failed.
        return -1
    }

    private func testURLCandidates(primary: String) -> [String] {
        var list: [String] = []
        func add(_ u: String) {
            let t = u.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !t.isEmpty, !list.contains(t) else { return }
            list.append(t)
        }
        // Prefer HTTP generate_204 (Clash / Stash / Shadowrocket style) — no TLS handshake noise.
        add(primary)
        add("http://www.gstatic.com/generate_204")
        add("http://cp.cloudflare.com/generate_204")
        add("http://www.msftconnecttest.com/connecttest.txt")
        return list
    }

    private func apiDelayOnce(
        proxyName: String,
        controller: String,
        secret: String,
        timeoutMs: Int,
        testURL: String
    ) async -> Int? {
        let encoded = Self.encodePath(proxyName)
        var components = URLComponents(string: "http://\(controller)/proxies/\(encoded)/delay")
        components?.queryItems = [
            URLQueryItem(name: "timeout", value: String(max(timeoutMs, 2000))),
            URLQueryItem(name: "url", value: testURL),
        ]
        guard let url = components?.url else { return nil }

        var request = URLRequest(url: url, timeoutInterval: TimeInterval(timeoutMs) / 1000.0 + 3.0)
        request.httpMethod = "GET"
        applyAuth(&request, secret: secret)

        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = TimeInterval(timeoutMs) / 1000.0 + 3
        config.waitsForConnectivity = false
        // Critical: API request itself must NOT go through system / mixed-port proxy.
        config.connectionProxyDictionary = [:]

        do {
            let (data, response) = try await URLSession(configuration: config).data(for: request)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return nil
            }
            if let message = json["message"] as? String, !message.isEmpty, json["delay"] == nil {
                return nil
            }
            return parseDelay(json)
        } catch {
            return nil
        }
    }

    private func parseDelay(_ json: [String: Any]) -> Int? {
        let raw: Int? = {
            if let v = json["delay"] as? Int { return v }
            if let v = json["delay"] as? Double { return Int(v.rounded()) }
            if let v = json["delay"] as? NSNumber { return v.intValue }
            return nil
        }()
        guard let raw, raw > 0, raw < 60_000 else { return nil }
        return raw
    }

    /// Fallback only: TCP connect RTT to node host:port (not through proxy tunnel).
    /// Resolves IPv4 first to avoid Happy Eyeballs ~1s IPv6→IPv4 stall that made every node look identical.
    private func tcpDelay(host: String, port: Int, timeoutMs: Int) async -> Int {
        guard !host.isEmpty, port > 0, port <= 65535 else { return -1 }

        let start = CFAbsoluteTimeGetCurrent()
        let endpointHost = await resolveIPv4Host(host) ?? host
        let connected = await connect(host: endpointHost, port: UInt16(port), timeoutMs: timeoutMs)
        guard connected else { return -1 }
        let ms = Int((CFAbsoluteTimeGetCurrent() - start) * 1000)
        return max(ms, 1)
    }

    private func resolveIPv4Host(_ host: String) async -> String? {
        if IPv4Address(host) != nil { return host }
        if IPv6Address(host) != nil { return nil }
        return await Task.detached(priority: .utility) {
            var hints = addrinfo(
                ai_flags: AI_ADDRCONFIG,
                ai_family: AF_INET,
                ai_socktype: SOCK_STREAM,
                ai_protocol: IPPROTO_TCP,
                ai_addrlen: 0,
                ai_canonname: nil,
                ai_addr: nil,
                ai_next: nil
            )
            var result: UnsafeMutablePointer<addrinfo>?
            defer { if result != nil { freeaddrinfo(result) } }
            guard getaddrinfo(host, nil, &hints, &result) == 0, let first = result else { return nil }
            var addr = first.pointee.ai_addr.withMemoryRebound(to: sockaddr_in.self, capacity: 1) { $0.pointee }
            var buf = [CChar](repeating: 0, count: Int(INET_ADDRSTRLEN))
            guard inet_ntop(AF_INET, &addr.sin_addr, &buf, socklen_t(INET_ADDRSTRLEN)) != nil else { return nil }
            return String(cString: buf)
        }.value
    }

    private func connect(host: String, port: UInt16, timeoutMs: Int) async -> Bool {
        await withCheckedContinuation { continuation in
            let queue = DispatchQueue(label: "bashx.speedtest")
            var tcp = NWProtocolTCP.Options()
            tcp.enableFastOpen = false
            let params = NWParameters(tls: nil, tcp: tcp)
            params.preferNoProxies = true
            let connection = NWConnection(
                host: NWEndpoint.Host(host),
                port: NWEndpoint.Port(rawValue: port)!,
                using: params
            )
            Self.startTCP(connection, queue: queue, timeoutMs: timeoutMs, continuation: continuation)
        }
    }

    private static func startTCP(
        _ connection: NWConnection,
        queue: DispatchQueue,
        timeoutMs: Int,
        continuation: CheckedContinuation<Bool, Never>
    ) {
        let lock = NSLock()
        var resumed = false
        let finish: @Sendable (Bool) -> Void = { ok in
            lock.lock()
            defer { lock.unlock() }
            guard !resumed else { return }
            resumed = true
            connection.cancel()
            continuation.resume(returning: ok)
        }

        connection.stateUpdateHandler = { state in
            switch state {
            case .ready:
                finish(true)
            case .failed, .cancelled:
                finish(false)
            default:
                break
            }
        }
        connection.start(queue: queue)

        queue.asyncAfter(deadline: .now() + .milliseconds(max(timeoutMs, 500))) {
            finish(false)
        }
    }

    private func applyAuth(_ req: inout URLRequest, secret: String) {
        let trimmed = secret.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmed.isEmpty {
            req.setValue("Bearer \(trimmed)", forHTTPHeaderField: "Authorization")
        }
    }

    private static func encodePath(_ value: String) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    private func intValue(_ value: Any?) -> Int? {
        if let v = value as? Int { return v }
        if let v = value as? Int64 { return Int(v) }
        if let v = value as? NSNumber { return v.intValue }
        if let v = value as? String { return Int(v) }
        return nil
    }
}

File: Android/app/src/main/java/com/bashx/app/net/SpeedTester.kt
package com.bashx.app.net

import com.bashx.app.clash.ClashConfigParser
import com.bashx.app.data.ProxyNode
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.net.Inet4Address
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Socket
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/**
 * Clash Verge / Stash style latency:
 * - Prefer mihomo `GET /proxies/{name}/delay` (URLTest through that node).
 * - TCP host:port only when controller is unavailable (reachability, not tunnel quality).
 */
object SpeedTester {
    data class Result(val name: String, val delayMs: Int)

    suspend fun testAll(
        nodes: List<ProxyNode>,
        timeoutMs: Int,
        concurrency: Int,
        controller: String? = null,
        secret: String = "",
        testURL: String,
        onProgress: suspend (String, Int) -> Unit,
    ): List<Result> = coroutineScope {
        val testables = nodes.filter { ClashConfigParser.isSpeedTestable(it) }
        if (testables.isEmpty()) return@coroutineScope emptyList()

        val apiTimeout = timeoutMs.coerceIn(2000, 8000)
        val limit = Semaphore(concurrency.coerceIn(1, 8))
        val catalog = if (controller != null) fetchProxyCatalog(controller, secret) else emptyMap()
        val urls = testURLCandidates(testURL)

        testables.map { node ->
            async(Dispatchers.IO) {
                limit.withPermit {
                    val delay = if (controller != null) {
                        apiDelay(node, controller, secret, apiTimeout, urls, catalog)
                    } else {
                        tcpDelay(node.server, node.port, apiTimeout)
                    }
                    onProgress(node.name, delay)
                    Result(node.name, delay)
                }
            }
        }.awaitAll()
    }

    private fun testURLCandidates(primary: String): List<String> {
        val out = linkedSetOf<String>()
        fun add(u: String) {
            val t = u.trim()
            if (t.isNotEmpty()) out += t
        }
        add(primary)
        add("http://www.gstatic.com/generate_204")
        add("http://cp.cloudflare.com/generate_204")
        add("http://www.msftconnecttest.com/connecttest.txt")
        return out.toList()
    }

    private fun fetchProxyCatalog(controller: String, secret: String): Map<String, String> {
        val client = apiClient(secret, 4000)
        val req = Request.Builder().url("http://$controller/proxies").build()
        val body = runCatching { client.newCall(req).execute().use { it.body?.string().orEmpty() } }
            .getOrDefault("")
        if (body.isBlank()) return emptyMap()
        val out = linkedMapOf<String, String>()
        runCatching {
            val proxies = JSONObject(body).optJSONObject("proxies") ?: return@runCatching
            for (key in proxies.keys()) {
                val obj = proxies.optJSONObject(key) ?: continue
                val type = obj.optString("type").lowercase()
                if (type in setOf("direct", "reject", "select", "url-test", "fallback", "load-balance", "relay", "pass", "compatible")) {
                    continue
                }
                out[key] = key
                val server = obj.optString("server")
                val port = obj.optInt("port", 0)
                if (server.isNotBlank() && port > 0) {
                    out["${server.lowercase()}:$port"] = key
                }
            }
        }
        return out
    }

    private fun resolveProxyName(node: ProxyNode, catalog: Map<String, String>): String {
        catalog[node.name]?.let { return it }
        return catalog["${node.server.lowercase()}:${node.port}"] ?: node.name
    }

    /** Fallback: TCP connect RTT. Prefer IPv4 to avoid Happy Eyeballs ~1s stall. */
    private fun tcpDelay(host: String, port: Int, timeoutMs: Int): Int {
        if (host.isBlank() || port <= 0) return -1
        val start = System.nanoTime()
        return try {
            val address = resolveIPv4(host) ?: InetAddress.getByName(host)
            Socket().use { socket ->
                socket.connect(InetSocketAddress(address, port), timeoutMs.coerceAtLeast(500))
            }
            ((System.nanoTime() - start) / 1_000_000L).toInt().coerceAtLeast(1)
        } catch (_: Exception) {
            -1
        }
    }

    private fun resolveIPv4(host: String): InetAddress? = runCatching {
        InetAddress.getAllByName(host).firstOrNull { it is Inet4Address }
    }.getOrNull()

    private fun apiDelay(
        node: ProxyNode,
        controller: String,
        secret: String,
        timeoutMs: Int,
        urls: List<String>,
        catalog: Map<String, String>,
    ): Int {
        val name = resolveProxyName(node, catalog)
        for (testURL in urls) {
            val ms = apiDelayOnce(name, controller, secret, timeoutMs, testURL)
            if (ms > 0) return ms
        }
        return -1
    }

    private fun apiDelayOnce(
        name: String,
        controller: String,
        secret: String,
        timeoutMs: Int,
        testURL: String,
    ): Int {
        val encoded = encodePath(name)
        val url = "http://$controller/proxies/$encoded/delay" +
            "?timeout=${timeoutMs.coerceAtLeast(2000)}&url=${URLEncoder.encode(testURL, Charsets.UTF_8.name())}"
        val client = apiClient(secret, timeoutMs + 3000)
        return runCatching {
            client.newCall(Request.Builder().url(url).get().build()).execute().use { resp ->
                if (!resp.isSuccessful) return@use -1
                val body = resp.body?.string().orEmpty()
                val json = runCatching { JSONObject(body) }.getOrNull() ?: return@use -1
                if (json.has("message") && !json.has("delay")) return@use -1
                val delay = json.optInt("delay", -1)
                if (delay <= 0 || delay >= 60_000) -1 else delay
            }
        }.getOrDefault(-1)
    }

    private fun encodePath(value: String): String {
        val allowed = (('a'..'z') + ('A'..'Z') + ('0'..'9') + "-._~").toSet()
        return buildString {
            for (ch in value) {
                if (ch in allowed) append(ch) else append(URLEncoder.encode(ch.toString(), Charsets.UTF_8.name()))
            }
        }
    }

    private fun apiClient(secret: String, timeoutMs: Int): OkHttpClient {
        val builder = OkHttpClient.Builder()
            .connectTimeout(timeoutMs.toLong(), TimeUnit.MILLISECONDS)
            .readTimeout(timeoutMs.toLong(), TimeUnit.MILLISECONDS)
            .proxy(java.net.Proxy.NO_PROXY) // API call must not go through system / mixed-port proxy
        if (secret.isNotBlank()) {
            builder.addInterceptor { chain ->
                chain.proceed(
                    chain.request().newBuilder()
                        .header("Authorization", "Bearer $secret")
                        .build()
                )
            }
        }
        return builder.build()
    }
}
